koku/api/provider/serializers.py

```diff
@@ -60,7 +63,61 @@ class Meta:
         """Metadata for the serializer."""
 
         model = ProviderBillingSource
-        fields = ('__all__')
+        fields = ('uuid', 'bucket')
+
+
+def _get_sts_access(provider_resource_name):
+    """Get for sts access."""
+    # create an STS client
+    sts_client = boto3.client('sts')
+
+    # Call the assume_role method of the STSConnection object and pass the role
+    # ARN and a role session name.
+    assumed_role = sts_client.assume_role(
+        RoleArn=provider_resource_name,
+        RoleSessionName='AccountCreationSession'
+    )
+    credentials = assumed_role.get('Credentials')
+    access_key_id = credentials.get('AccessKeyId') if credentials else None
+    secret_access_key = credentials.get('SecretAccessKey') if credentials else None
+    session_token = credentials.get('SessionToken') if credentials else None
+
+    return (access_key_id, secret_access_key, session_token)
+
+
+def _check_s3_access(access_key_id, secret_access_key,
+                     session_token, bucket):
+    """Check for access to s3 bucket."""
+    s3_exists = True
+    s3_resource = boto3.resource(
+        's3',
+        aws_access_key_id=access_key_id,
+        aws_secret_access_key=secret_access_key,
+        aws_session_token=session_token,
+    )
+    try:
+        s3_resource.meta.client.head_bucket(Bucket=bucket)
```

Comment:
Since we're only interacting with the client, why not just make an s3 client, instead of the resource?

Reply:
I think the con of using the assumed role is that we either need to grant broad access to all S3 buckets in our role JSON, or have the customer fill in their bucket name into the role JSON to restrict access down to that specific bucket. That becomes a little less copy-paste.

To me the pro of using our account is that once a customer is on-boarded, the bulk of our interaction with their AWS account is just pulling the reports from S3. S3 makes it really easy to provide bucket-level access to other accounts, so for data ingestion, if we're using our account, we wouldn't have to worry about the assume_role step, and the permissions are easily locked to a single bucket.


https://docs.aws.amazon.com/AmazonS3/latest/dev/s3-access-control.html talks about resource based and user based access policies. With resource based you can give another AWS account access to some or all actions on a single bucket.